### PR TITLE
feat: autoscaling for MonoVertex

### PR DESCRIPTION
**Pull request by whynowy**
Explain what this PR does.

### CODE REVIEW

File: pkg/mvtxdaemon/server/service/mvtx_service.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+// getPending returns the pending count for the mono vertex
+func (mvs *MoveVertexService) getPending(ctx context.Context) map[string]*wrapperspb.Int64Value {
+	log := logging.FromContext(ctx)
+	headlessServiceName := mvs.monoVtx.GetHeadlessServiceName()
+	pendingMap := make(map[string]*wrapperspb.Int64Value)
+
+	// Get the headless service name
+	// We can query the metrics endpoint of the (i)th pod to obtain this value.
+	// example for 0th pod : https://simple-mono-vertex-mv-0.simple-mono-vertex-mv-headless:2469/metrics
+	url := fmt.Sprintf("https://%s-mv-1.%s.%s.svc:%v/metrics", mvs.monoVtx.Name, headlessServiceName, mvs.monoVtx.Namespace, v1alpha1.MonoVertexMetricsPort)
```
Comment (kohlisid) [inline pkg/mvtxdaemon/server/service/mvtx_service.go:82]: Is this URL for a specific pod?
Or the MonoVertex svc
```
simple-mono-vertex-mv-headless     ClusterIP   None             <none>        2469/TCP   27h
```

Comment (whynowy) [reply pkg/mvtxdaemon/server/service/mvtx_service.go:82]: Yes, for pending, one pod is good enough.

Comment (whynowy) [reply pkg/mvtxdaemon/server/service/mvtx_service.go:82]: Make it `...-mv-0...` please, @kohlisid

File: server/apis/v1/handler.go [hunk 2]
```diff
@@ -118,21 +120,25 @@ func NewHandler(ctx context.Context, dexObj *DexObject, localUsersAuthObject *Lo
...
-		healthChecker:        NewHealthChecker(ctx),
-		opts:                 o,
+		kubeClient:            kubeClient,
+		metricsClient:         metricsClient,
+		numaflowClient:        numaflowClient,
+		daemonClientsCache:    daemonClientsCache,
+		mvtDaemonClientsCache: mvtDaemonClientsCache,
+		dexObj:                dexObj,
+		localUsersAuthObject:  localUsersAuthObject,
+		healthChecker:         NewHealthChecker(ctx),
```
Comment (kohlisid) [inline server/apis/v1/handler.go:140]: We have not configured the health checker for monovertex. I believe we can remove for now and we can add back once I add support for this

Comment (KeranYang) [reply server/apis/v1/handler.go:140]: I haven't used healthChecker for mono vertex APIs. Currently, the health status for mono vertex is hardcoded as healthy. We can keep healthChecker and work on expanding it to mono vertex.

File: server/apis/v1/handler.go [hunk 3]
```diff
@@ -1335,3 +1362,23 @@ func (h *handler) getDaemonClient(ns, pipeline string) (daemonclient.DaemonClien
 		return dClient, nil
 	}
 }
+
+func (h *handler) getMonoVertexDaemonClient(ns, mvtName string) (mvtdaemonclient.MonoVertexDaemonClient, error) {
+	if mvtDaemonClient, ok := h.mvtDaemonClientsCache.Get(daemonSvcAddress(ns, mvtName)); !ok {
```
Comment (whynowy) [inline server/apis/v1/handler.go:1367]: should have a different function for mvtx daemon svc address, they are different.

Comment (KeranYang) [reply server/apis/v1/handler.go]: thanks, addressed.

File: server/apis/v1/handler.go [hunk 4]
```diff
@@ -1339,44 +1339,52 @@ func validatePipelinePatch(patch []byte) error {
...
+func pipelineDaemonSvcAddress(ns, pipelineName string) string {
+	// the format is consistent with what we defined in GetDaemonServiceURL in `pkg/apis/numaflow/v1alpha1/pipeline_types.go`
+	// do not change it without changing the other.
+	return fmt.Sprintf("%s.%s.svc:%d", fmt.Sprintf("%s-daemon-svc", pipelineName), ns, dfv1.DaemonServicePort)
+}
+
+func monoVertexDaemonSvcAddress(ns, monoVertexName string) string {
+	// the format is consistent with what we defined in GetDaemonServiceURL in `pkg/apis/numaflow/v1alpha1/mono_vertex_types.go`
+	// do not change it without changing the other.
+	return fmt.Sprintf("%s.%s.svc:%d", fmt.Sprintf("%s-mv-daemon-svc", monoVertexName), ns, dfv1.DaemonServicePort)
```
Comment (whynowy) [inline server/apis/v1/handler.go:1351]: port.

Comment (whynowy) [reply server/apis/v1/handler.go:1351]: @KeranYang

Comment (KeranYang) [reply server/apis/v1/handler.go]: sorry, addressed.

**Review by vigith — APPROVED** [review]

**Pull request merged**